Refactor Building.Footprint, replace FootprintUtils, add untargetable cells

reaperrr opened the pull request:
Refactors `Building.Footprint` loading and replaces `FootprintUtils` with BuildingInfo equivalents.

Adds a 'u' letter to Footprint that behaves like 'x', except that it is ignored by `HitShape.UseTargetableCellsOffsets = true`.

While the amount of yaml lines and custom `TargetableOffsets` it saves in the shipping mods might fall into the *'nice, but is that really worth it?'* category, the savings in TS would be tremendous, so much that I'd refuse setting up custom `TargetableOffsets` in TS and just stick to occupied cells if this were rejected.

Applied it to all shipping mods (TS will be adressed with #13464).

Depends on #13561.
Closes #13476.
Dependency for doing #13464 properly.

reaperrr committed:
Simplify D2k TargetableOffsets rules

By using the new 'u' Footprint letter.

reaperrr committed:
Simplify TD base building hitshape rules

reaperrr committed:
Simplify RA base building hitshape rules

rob-v commented on OpenRA.Mods.Common/Traits/Buildings/FootprintUtils.cs:
```diff
@@ -43,13 +43,20 @@ public static IEnumerable<CPos> FrozenUnderFogTiles(Actor a)
 
+		public static IEnumerable<CPos> UnpathableTiles(string name, BuildingInfo buildingInfo, CPos position)
+		{
+			var footprint = buildingInfo.Footprint.Where(x => !char.IsWhiteSpace(x)).ToArray();
```
`buildingInfo.Footprint.Where(x => !char.IsWhiteSpace(x))`
is now used on 5 places - small suggestion - to cache it for little shorter/faster code.

pchote replied:
At this stage it would be a lot cleaner to replace all this DIY parsing with a custom `LoadUsing` that stores `buildingInfo.Footprint` as a 2D array of an enum describing these different types.

reaperrr replied:
I don't know how the code for that would have to look like... do we have something comparable that I could take as reference?

pchote replied:
```csharp
public enum FootprintCellType { Passable, Blocking, ... }
static readonly Dictionary<char, FootprintCellType> FootprintCellTypes = new Dictionary<char, FootprintCellType>()
{
	{ '_', FootprintCellType.Passable },
	{ 'x', FootprintCellType.Blocking },
	...
};
```
```csharp
[FieldLoader.LoadUsing("LoadFootprint")]
public readonly Dictionary<CVec, FootprintCellType> Footprint;

protected static object LoadFootprint(MiniYaml yaml)
{
	var footprintYaml = yaml.Nodes.FirstOrDefault(n => n.Key == "Footprint");
	var footprintChars = footprintYaml != null ? footprintYaml.Value.Value.Where(x => !char.IsWhiteSpace(x)).ToArray() : new[] { 'x' };

	var dimensionsYaml = yaml.Nodes.FirstOrDefault(n => n.Key == "Dimensions");
	var dim = dimensionsYaml != null ? FieldLoader.GetValue<CVec>("Dimensions", dimensionsYaml.Value.Value) : new CVec(1, 1);

	var index = 0;
	var ret = new Dictionary<CVec, FootprintCellType>();
	for (var y = 0; y < dim.Y; y++)
	{
		for (var x = 0; x < dim.X; x++)
		{
			FootprintCellType cell;
			if (!FootprintCellTypes.TryGetValue(footprintChars[index++], out cell))
				throw new YamlException("Invalid footprint cell type '{0}'".F(footprintChars[index - 1]));

			ret[new CVec(x, y)] = cell;
		}
	}

	if (index != dim.X * dim.Y)
		throw new YamlException("Invalid footprint");

	return ret;
}

public IEnumerable<CPos> UnpathableTiles(CPos location)
{
	return NewFootprint.Where(kv => kv.Value == BuildingInfo.FootprintCellType.Blocking).Select(kv => location + kv.Key);
}

...
```
and then you can remove `FootprintUtils` and query directly from the `BuildingInfo`.

rob-v replied:
sorry @reaperrr for making small suggestion...

pchote replied:
I just simplified my example a bit, so check again if you have already started on this.

reaperrr replied:
@pchote I'm struggling with how to look up and include Bibs.

As for the rest, I can probably handle it...

pchote replied:
Bibs have been an endless PITA.  My preferred solution would be to make `Bib` a visual-only trait (`WithBuildingBib` in `Mods.Cnc`?) and require mods to define non-mini bibs as part of the footprint using your new `u` tiles.  We then just need a `WVec` offset for the `CenterPosition` to avoid moving it down by half a cell.

rob-v commented on mods/d2k/rules/structures.yaml:
```diff
@@ -599,7 +598,7 @@ starport:
 		Cost: 1500
 	Building:
-		Footprint: xxx x=x =x=
+		Footprint: xxx xxx =x=
```
It seems better not to move units into middle of startport, but the commit comment 'allowed both friendly and enemy actors to block the exit' seems incorrect - in current release unit in middle doesn't prevent leaving new units according to my test.

rob-v commented on mods/d2k/rules/structures.yaml:
```diff
@@ -883,14 +882,13 @@ high_tech_factory:
 		ExitCell: 0,0
 	Building:
-		Footprint: _x_ xxx xxx
+		Footprint: _u_ xxx uuu
```
Surely fine, the first 'top u' seems unnecessary, as there is HitShape@TOP.

reaperrr replied:
Both this and the Research Center define a custom, slightly lowered offset for the top cell, so this needs to be `u` otherwise the main HitShape would add another targetable pos at the cell center.

rob-v commented on mods/d2k/rules/structures.yaml:
```diff
@@ -961,14 +959,13 @@ research_centre:
 		Name: IX Research Center
 	Building:
-		Footprint: _x_ xxx xxx
+		Footprint: _u_ xxx uuu
```
x instead of 'top u'?

rob-v reviewed (commented):
The Building.Footprint description can be updated with 'u' and also '='. All great, only small questions.

reaperrr committed:
Refactor footprint cell lookups and move them to Building

Removing FootprintUtils happens in the next commit for better
reviewability.

reaperrr committed:
Remove FootprintUtils

reaperrr commented:
I'm done with implementing @pchote's suggested code changes locally, just waiting for #13561 to be reviewed and merged now.

reaperrr committed:
Add OccupiedUntargetable FootprintCellType

For cells that are occupied & unpassable, but should be ignored by HitShape.UseOccupiedCellsOffsets.

reaperrr committed:
Rename UseOccupiedCellsOffsets to UseTargetableCellsOffsets

With the introduction of the new 'u' BlockingUntargetable cell type, the old property name became inaccurate.

reaperrr commented on mods/d2k/rules/structures.yaml:
```diff
@@ -272,7 +272,6 @@ refinery:
 			TopLeft: -512, -1024
 			BottomRight: 1536, 0
 	HitShape@BOTTOM:
-		UseOccupiedCellsOffsets: false
```
Note to reviewers:  When originally setting this up, I simply forgot that 'false' is actually the internal default, and therefore this line is redundant for all shapes that are not inherited (i.e. all `@FOO` shapes). Applies to all 3 shipping mods.

reaperrr commented:
Updated. Still depends on #13561, but I pushed the following changes so you can start some preliminary reviewing if you feel like it.

The first two commits are the tricky ones in terms of code difficulty, the other 5 commits are pretty trivial, just a bit tedious to go through.

reaperrr commented:
Rebased.
github messed up the commit order a bit, yaml commits should be 4-6, but that shouldn't really matter for reviewing.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -58,9 +68,97 @@ public class BuildingInfo : ITraitInfo, IOccupySpaceInfo, IPlaceBuildingDecorati
+				yield return t;
+		}
+
+		public IEnumerable<CPos> Tiles(Actor actor)
```
This is only used from a handful of places, so it would be better to remove this and have them call `Tiles(CPos)` directly.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -23,6 +23,14 @@ namespace OpenRA.Mods.Common.Traits
+		Empty = '_',
+		OccupiedPassable = '=',
+		Blocking = 'x',
+		BlockingUntargetable = 'u'
```
How about `X` instead of `u`?  This would be more intuitive IMO as this is only a minor modification of the normal `x`.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -23,6 +23,14 @@ namespace OpenRA.Mods.Common.Traits
+	{
+		Empty = '_',
+		OccupiedPassable = '=',
+		Blocking = 'x',
```
Maybe rename this to `Occupied` for consistency with `=`?

pchote reviewed (approved):
Looks good.  Just a few style suggestions, and needs another rebase.

reaperrr commented:
Rebased and updated.

reaperrr commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -23,6 +23,13 @@ namespace OpenRA.Mods.Common.Traits
 	public class GivesBuildableAreaInfo : TraitInfo<GivesBuildableArea> { }
 	public class GivesBuildableArea { }
 
+	public enum FootprintCellType
```
Fixed.

ltem commented:
I found a small regression from #13561. The palace structure in d2k is a 3x3 building with a "bib" line at the bottom row. Right now it is a 3x4 building with a bib line like the heavy factory.

([Line 1018](https://github.com/OpenRA/OpenRA/pull/13553/commits/9c44ce241741e705f93097cc4729b1200d1bad04#diff-76baa349824e533ba919e9780fa23362R1018))


|Screenshot from Gruntmods|
|:---:|
|![screenie](https://user-images.githubusercontent.com/20945684/27997864-3dfe1b72-6501-11e7-82fc-2c904fd95172.png)|

reaperrr committed:
Fix D2k Palace footprint

reaperrr commented:
> I found a small regression from #13561. The palace structure in d2k is a 3x3 building with a "bib" line at the bottom row. Right now it is a 3x4 building with a bib line like the heavy factory.

Good catch. Fixed.

abcdefg30 commented:
👍 / ✅

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c9d836fc640a63777a438edfbf1f633ef1c0fd9d...0a4ac6bb1981753dbbf3e017ae6a51ded1a29061)
